src/librustc_mir/build/scope.rs

```diff
@@ -800,7 +800,7 @@ fn build_free<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
                               -> TerminatorKind<'tcx> {
     let free_func = tcx.lang_items.require(lang_items::BoxFreeFnLangItem)
                        .unwrap_or_else(|e| tcx.sess.fatal(&e));
-    let substs = Substs::new(tcx, iter::once(Kind::from(data.item_ty)));
+    let substs = tcx.mk_substs(iter::once(Kind::from(data.item_ty)));
```

Comment:
I feel like this might be easier to optimize if we skip the `iter::once`, since we're not joining it with anything anyway. Leaving this is a note to self, mostly, since I'm not sure we *can* omit the iterator with the current implementation.